test/Function/StackArgsWithFormals.js

```diff
@@ -0,0 +1,141 @@
+//Test1 : Tests Optimization with bailout.
+var shouldBailout = false;
+var test1 = function (argMath3) 
+{
+    if(shouldBailout)
+    {
+        arguments[0] = 4;
+        for(var o in arguments)
+        {
+            print(o);
+        }
+    }
+    WScript.Echo(argMath3);
+};
+
+print("TEST 1 : ");
+test1(10,2); // Interpreter
+shouldBailout = true;
+test1(10,2); // full jit
+
+//Test2: Test Inlinee Arguments optimization with StackArgsWithFormals
+var inner_test2 = function(argMath6=1) {
+    return argMath6;  
+};
+
+function test2(a,b,c) {
+  inner_test2.apply({}, arguments);
+}
+print("TEST 2 : ");
+print(test2(1,2,3));
+print(test2(4,5,6));
+
+//Test3: Non Simple Parameter List
+
+function test3(a, b =2, c = 5)
+{
+    if(shouldBailout)
+    {
+        arguments[2] = 10;
+        var result = arguments[2] + a;
+        print(result);
+        return;
+    }
+    print(a+b+arguments[1]);
+}
+
+print("TEST 3 : ");
+shouldBailout = false;
+test3(1,2,3,4);
+test3(5,6,7,8);
+
+shouldBailout = true;
+test3(10,11,12,14);
+
+
+//Test4: With Rest argument
+
+function test4(a,b,c,d,e) {
+  inner_test4(a,b,c,d);
+}
+
+var inner_test4 = function (a, b, ...argArr0) {
+    if (arguments.length > 1) {
+        print(a + b);
+        print(argArr0[0])
+    }
+  };
+  
+print("TEST 4 : ");
+test4(1,2,3,4);
+test4(5,6,7,8);
+
+
+//Test5: Inside Loop and Bails out due to out of range of actuals access
+function test5(a,b,c=10)
+{
+    var result = 0;
+    for(var i = 0; i < 5; i++)
+    {
+        result += arguments[i];
+    }
+    result += (a+c);
+    print(result);
+}
+
+print("TEST 5 : ");
+test5(1,2,3,4,5,6);
+test5(1,2,3);// Bail out here.
+
+//Test6 : Using strict mode
+
+function test6(a,b,c)
+{
+    'use strict';
+    print(arguments[0] + arguments[2] + a + b);
+}
+
+print("TEST 6 : ");
+test6(1,2,3);
+test6(4,5,6,7);
+
+//Test7 : Use strict with bail out
+
+function test7(a,b,c)
+{
+    'use strict';
+    if(shouldBailout)
+    {
+        arguments[0] = 4;
+    }
+    print(a);
+}
+
+print("TEST 7 : ");
+shouldBailout = false;
+test7();
+shouldBailout = true;
+test7();
+
+//Test 8 : for-in statement
+function test8(a,b)
+{
+	for(a in [10,20,30])
+	{
+		print(a);
+		print(arguments[0]);
+	}
+}
+print("TEST 8 : ");
+test8(1,2);
+test8('x','y');
+
+//Test 9 : Destructuring.
+function test9(a,b)
+{
+	[a,b] = [10,20];

```

Comment:
Done.

---
In reply to: [67766996](https://github.com/Microsoft/ChakraCore/pull/1135#discussion_r67766996) [](ancestors = 67766996)